lib/pbench/test/unit/agent/test_config.py

```diff
@@ -34,3 +38,72 @@ def test_results_attr():
         "pbench"
         in PbenchAgentConfig(os.environ["_PBENCH_AGENT_CONFIG"]).results["user"]
     )
+
+
+# Monkeypatching different Classes and Methods to test Pbench-run value
+
+
+class MockPbenchAgentConfig:
+    """Mock class for PbenchAgentConfig"""
+
+    def __init__(self, cfg_name: str):
+        self.pbench_run = "/test/pbench-run"
+        self.pbench_log = "logger"
+        self.pbench_install_dir = Path("/test/install_dir")
+        self.ssh_opts = "/ssh_opts"
+        self.scp_opts = "/scp_opts"
+
+
+def mock_exist(path: Path):
+    """mock method for Path.exist"""
+    return True
+
+
+def mock_current_context():
+    """mock get_current_context from click package"""
+    None
+
+
+class MockContext:
+    """Mocking Context"""
+
+    config = "/test/pbench-config-path"
+
+
+class TestCommand(BaseCommand):
+    """Mock of class that extends BaseCommand"""
+
+    def __init__(self):
+        super(TestCommand, self).__init__(MockContext())
+
+    def execute(self):
+        pass
+
+
+def test_pbench_run_path(monkeypatch):
+    """Use pbench_run value from pbench-agent.cfg file"""
+    assert os.environ.get("pbench_run") is None
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(Path, "exists", mock_exist)
+    pbench_run_val = TestCommand().pbench_run
+    assert str(pbench_run_val) == "/test/pbench-run"
+
+
+def test_pbench_run_as_env(monkeypatch):
+    """Use pbench_run value, passed manually through
+    `pbench-run` environment variable"""
+    monkeypatch.setenv("pbench_run", "/test/pbench-run/environ")
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(Path, "exists", mock_exist)
+    pbench_run_val = TestCommand().pbench_run
+    assert str(pbench_run_val) == "/test/pbench-run/environ"
+
+
+def test_pbench_run_dir_existence(monkeypatch):
+    """Verify that pbench_list_tools fail when pbench_run is defined
+    to a directory that doesn't exist"""
+    monkeypatch.setattr(PbenchAgentConfig, "__init__", MockPbenchAgentConfig.__init__)
+    monkeypatch.setattr(click, "get_current_context", mock_current_context)
```

Comment:
I suspect that we want a slightly different approach, here.  We _could_ mock `get_current_context()`, but we would need the mock function to actually return a usable `click.Context` object (although, we could probably mock the object fairly easily).  Instead, I think you can mock just the `exit()` function on that class.

I would recommend having the mock `exit()` function raise a custom exception and have the test catch it.  You can include the exit status as an argument to the exception and check it when it's caught.

Reply:
yes, I can understand, I was facing some issues here. I also tried to mock the exit(), which module does the `exit()` belong to? Is it `sys`? I tried with `click`& I also tried with `sys` but none of them is working.

Reply:
_If_ you need to mock it (which I don't think you do, if you use the Click test runner), it would be something close to
```
    monkeypatch.setattr(click.Context, "exit", lambda s: raise MyExitException(s))
```
where `MyExitException` is a class that you define which captures the exit status and makes it available to the code which catches the exception.